src/github.com/matrix-org/dendrite/cmd/mediaapi-integration-tests/main.go

```diff
@@ -0,0 +1,178 @@
+- width: 32
+  height: 32
+  method: crop
+- width: 96
+  height: 96
+  method: crop
+- width: 320
+  height: 240
+  method: scale
+- width: 640
+  height: 480
+  method: scale
+- width: 800
+  height: 600
+  method: scale
+`)
+
+const serverType = "media-api"
+
+var testDatabaseTemplate = "dbname=%s sslmode=disable binary_parameters=yes"
+
+var timeout time.Duration
+
+func startMediaAPI(suffix string, dynamicThumbnails bool) (*exec.Cmd, chan error, string) {
+	dir, err := ioutil.TempDir("", serverType+"-server-test"+suffix)
+	if err != nil {
+		panic(err)
+	}
+
+	configFilename := serverType + "-server-test-config" + suffix + ".yaml"
+	configFileContents := makeConfig(suffix, dir, dynamicThumbnails)
+
+	serverArgs := []string{
+		"--config", configFilename,
+		"--listen", "localhost:1777" + suffix,
+	}
+
+	databases := []string{
+		testDatabaseName + suffix,
+	}
+
+	cmd, cmdChan := test.StartServer(
+		serverType,
+		serverArgs,
+		suffix,
+		configFilename,
+		configFileContents,
+		postgresDatabase,
+		postgresContainerName,
+		databases,
+	)
+	return cmd, cmdChan, dir
+}
+
+func makeConfig(suffix, basePath string, dynamicThumbnails bool) string {
+	return fmt.Sprintf(
+		`
+server_name: "%s"
+base_path: %s
+max_file_size_bytes: %s
+database: "%s"
+dynamic_thumbnails: %s
+%s`,
+		"localhost:1777"+suffix,
+		basePath,
+		"10485760",
+		fmt.Sprintf(testDatabaseTemplate, testDatabaseName+suffix),
+		strconv.FormatBool(dynamicThumbnails),
+		thumbnailPregenerationConfig,
+	)
+}
+
+func cleanUpServer(cmd *exec.Cmd, dir string) {
+	cmd.Process.Kill() // ensure server is dead, only cleaning up so don't care about errors this returns.
+	if err := os.RemoveAll(dir); err != nil {
+		fmt.Printf("WARNING: Failed to remove temporary directory %v: %q\n", dir, err)
+	}
+}
+
+// Runs a battery of media API server tests
+// The tests will pause at various points in this list to conduct tests on the HTTP responses before continuing.
+func main() {
+	fmt.Println("==TESTING==", os.Args[0])
+
+	var err error
+	timeout, err = time.ParseDuration(timeoutString)
+	if err != nil {
+		fmt.Printf("ERROR: Invalid timeout string %v: %q\n", timeoutString, err)
+		return
+	}
+
+	// create server1 with only pre-generated thumbnails allowed
+	server1Cmd, server1CmdChan, server1Dir := startMediaAPI("1", false)
+	defer cleanUpServer(server1Cmd, server1Dir)
+	testDownload("1", "localhost:17771", "doesnotexist", "", 404, server1CmdChan)
+
+	// create server2 with dynamic thumbnail generation
+	server2Cmd, server2CmdChan, server2Dir := startMediaAPI("2", true)
+	defer cleanUpServer(server2Cmd, server2Dir)
+	testDownload("2", "localhost:17772", "doesnotexist", "", 404, server2CmdChan)
+}
+
+func getMediaURI(scheme, host, endpoint string, components []string) string {
+	pathComponents := []string{
+		host,
+		"api",
+		"_matrix",
+		"media",
+		"v1",
+		endpoint,
+	}
+	pathComponents = append(pathComponents, components...)
+	return scheme + path.Join(pathComponents...)
+}
```

Comment:
I wonder if this would be nicer written as:

```go
func getMediaURI(scheme, host, endpoint string, components ...string) string {
 	pathComponents := []string{host,"/api/_matrix/media/v1", endpoint}
 	pathComponents = append(pathComponents, components...)
 	return scheme + path.Join(pathComponents...)
 }
```